packages/table/src/common/batcher.ts

```diff
@@ -0,0 +1,220 @@
+ *
+ *     batcher.startNewBatch();
+ *
+ *     allChildrenKeys.forEach((prop1: string, index: number) => {
+ *         batcher.addArgsToBatch(prop1, "prop2", index);
+ *     });
+ *
+ *     batcher.removeOldAddNew((prop1: string, prop2: string, other: number) => {
+ *         return <Child prop1={prop1} prop2={prop2} other={other} />;
+ *     });
+ *
+ *     if (!batcher.isDone()) {
+ *         batcher.idleCallback(this.forceUpdate());
+ *     }
+ *
+ *     const currentChildren = batcher.getList();
+ *     return currentChildren;
+ * }
+ *
+ * ```
+ */
+export class Batcher<T> {
+    public static DEFAULT_ADD_LIMIT = 20;
+    public static DEFAULT_UPDATE_LIMIT = 20;
+    public static DEFAULT_REMOVE_LIMIT = 20;
+
+    private currentObjects: Record<string, T> = {};
+    private oldObjects: Record<string, T> = {};
+    private batchArgs: Record<string, any[]> = {};
+    private done = true;
+    private callback: Callback;
+
+    /**
+     * Resets the "batch" and "current" sets. This essentially clears the cache
+     * and prevents accidental re-use of "current" objects.
+     */
+    public reset() {
+        this.batchArgs = {};
+        this.oldObjects = this.currentObjects;
+        this.currentObjects = {};
+    }
+
+    /**
+     * Starts a new "batch" argument set
+     */
+    public startNewBatch() {
+        this.batchArgs = {};
+    }
+
+    /**
+     * Stores the variadic arguments to be later batched together.
+     *
+     * The arguments must be simple stringifyable objects.
+     */
+    public addArgsToBatch(...args: SimpleStringifyable[]) {
+        this.batchArgs[args.join("|")] = args;
+    }
+
+    /**
+     * Compares the set of "batch" arguments to the "current" set. Creates any
+     * new objects using the callback as a factory. Removes old objects.
+     *
+     * Arguments that are in the "current" set but were not part of the last
+     * "batch" set are considered candidates for removal. Similarly, Arguments
+     * that are part of the "batch" set but not the "current" set are candidates
+     * for addition.
+     *
+     * The number of objects added and removed may be limited with the
+     * `...Limit` parameters.
+     *
+     * Finally, the batcher determines if the batching is complete if the
+     * "current" arguments match the "batch" arguments.
+     */
+    public removeOldAddNew(
+            callback: (...args: any[]) => T,
+            addNewLimit = Batcher.DEFAULT_ADD_LIMIT,
+            removeOldLimit = Batcher.DEFAULT_REMOVE_LIMIT,
+            updateLimit = Batcher.DEFAULT_UPDATE_LIMIT,
+    ) {
+
+        // remove old
+        const toRemove = this.setKeysOperation(this.currentObjects, this.batchArgs, "difference", removeOldLimit);
+        toRemove.forEach((key) => delete this.currentObjects[key]);
+
+        // remove ALL old object not in batch
+        const toRemoveOld = this.setKeysOperation(this.oldObjects, this.batchArgs, "difference", -1);
```

Comment:
i disagree. I think it's more natural (or at least more common) to use a number than a `null` value